docs/book/src/reference/using-finalizers.md

```diff
@@ -0,0 +1,99 @@
+# Using Finalizers
+
+`Finalizers` allow controllers to implement asynchronous pre-delete hooks. Let
+say, you create an external resource such as a storage bucket for each object of
+an API type you are implementing and you need to delete the associated external resource
+when the corresponding object is deleted from Kubernetes, you can use a
+finalizer to delete the external resource.
+
+You can read more about the finalizers in the [kubernetes reference docs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers). Section below
+demonstrates how to register and trigger pre-delete hooks in the `Reconcile`
+method of a controller.
+
+Highlights:
+- If object is not being deleted and does not have the finalizer registered,
+  then add the finalizer and update the object in kubernetes.
+- If object is being deleted and the finalizer is still present in finalizers list,
+  then execute the pre-delete logic and remove the finalizer and update the
+  object.
+- You should implement the pre-delete logic in such a way that it is safe to 
+ invoke it multiple times for the same object.
+
+```go
```

Comment:
use `literatego` for this -- you can put the full code in, but mark sections that you don't want to emphasize as collapsed (see `main_revisited.md`)

Reply:
Thanks for the tip, exploring it.

Reply:
I took a look. Earlier I was thinking of reusing Cronjob controller file to demonstrate that, but looks like sharing the same `.go` file for two different pages will be challenging because the two pages will have different requirement for collapsing portions of the `.go` file.

I took a look at `empty_controller.go` also, so I can follow that pattern also where I write this in another `.go` file, but then I can't use `cronjob` reconciler because it won't compile (not sure if we compile the Go files that we have under docs) because I will end-up overriding Reconciler. 

I am trying to understand now which approach is better ? (code embedded in markdowns or markdown (HTML) embedded in Go) :)? 

Want to understand how are you approaching this ?

Reply:
For now, I have created a file `finalizer_example.go` and moved the go code there so that we have something consistent to begin with. We don't sanity check (compile) go files under book dir but I see a few challenges around that.